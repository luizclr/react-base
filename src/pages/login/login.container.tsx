import { useEffect } from "react";
import { useNavigate } from "react-router-dom";

import { User } from "~/entities/user";
import { Login as BaseLogin } from "~/pages/login/login";
import { PATHS } from "~/routes/paths";
import { useApp } from "~/state/app/hook";
import { useAuth } from "~/state/auth/hook";

export const Login: React.FC = () => {
  const navigate = useNavigate();
  const { setIsLoading, storageService } = useApp();
  const { signIn, isAuthenticated, authService } = useAuth();

  useEffect((): void => {
    if (isAuthenticated) navigate(PATHS.HOME);
  }, [isAuthenticated]);

  const handleSubmit = async (email: string, password: string): Promise<void> => {
    setIsLoading(true);

    await authService.authenticate({ email, password }, { onSuccess, onUnauthorized, onError });
  };

  const onSuccess = (token: string, user: User): void => {
    storageService.set("token", token);
    storageService.set("user", JSON.stringify(user));

    signIn(user, token);

    setIsLoading(false);
    navigate(PATHS.HOME);
  };

  const onUnauthorized = (): void => {
    setIsLoading(false);
  };

  const onError = (): void => {
    setIsLoading(false);
  };

  return <BaseLogin handleSubmit={handleSubmit} />;
};
