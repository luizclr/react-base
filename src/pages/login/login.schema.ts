import { z } from "zod";

export const LoginSchema = z.object({
  email: z.string().email().nonempty(),
  password: z
    .string()
    .min(8, { message: "Password must contain at least 8 characters" })
    .nonempty(),
});

export type LoginSchemaType = z.infer<typeof LoginSchema>;
