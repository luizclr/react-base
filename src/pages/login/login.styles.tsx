import { getColor, getFontSize, getFontWeight, getSize } from "react-styled-guide";
import styled from "styled-components";

export const Container = styled.div`
  background-color: ${getColor(({ success }) => success.medium)};
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
  padding: ${getSize(({ xxxs }) => xxxs)};
  color: ${getColor(({ neutral }) => neutral.dark)};
`;

export const Card = styled.div`
  background-color: #fff;
  border-radius: ${getSize(({ nano }) => nano)};
  padding: ${getSize(({ xxs }) => xxs)};
  max-width: 100%;
  width: 400px;
`;

export const Title = styled.h2`
  font-size: ${getFontSize(({ m }) => m)};
  margin-bottom: ${getSize(({ xxs }) => xxs)};
  color: ${getColor(({ neutral }) => neutral.dark)};
`;

export const ButtonWrapper = styled.div`
  display: flex;
  justify-content: center;
  background-color: inherit;
`;

export const Button = styled.button`
  background-color: ${getColor(({ secondary }) => secondary.medium)};
  border-radius: ${getSize(({ quark }) => quark)};
  padding: ${getSize(({ nano }) => nano)};
  border: none;
  color: #fff;
  font-weight: ${getFontWeight(({ bold }) => bold)};
  margin: 0 auto;
  margin-top: ${getSize(({ nano }) => nano)};
  width: 50%;

  :hover {
    background-color: ${getColor(({ secondary }) => secondary.tinyDark)};
    cursor: pointer;
  }
`;
