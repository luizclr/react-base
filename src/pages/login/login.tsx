import { zodResolver } from "@hookform/resolvers/zod";
import { FieldValues, useForm } from "react-hook-form";

import { InputFIeld } from "~/components/input-field/input-field";
import { LoginSchema } from "~/pages/login/login.schema";
import { Button, ButtonWrapper, Card, Container, Title } from "~/pages/login/login.styles";

type Props = {
  handleSubmit: (email: string, password: string) => Promise<void>;
};

export const Login: React.FC<Props> = (props: Props) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm({
    resolver: zodResolver(LoginSchema),
  });

  const onSubmit = async ({ email, password }: FieldValues): Promise<void> => {
    await props.handleSubmit(email, password);
  };

  return (
    <Container>
      <Card>
        <form onSubmit={handleSubmit(onSubmit)}>
          <Title>Login</Title>
          <InputFIeld
            type="email"
            name="email"
            label="E-mail"
            register={register}
            placeholder="example@example.com"
            errors={errors}
          />
          <InputFIeld
            type="password"
            name="password"
            label="Password"
            register={register}
            placeholder="**********"
            errors={errors}
          />
          <ButtonWrapper>
            <Button data-testid="submit" type="submit">
              SIGN IN
            </Button>
          </ButtonWrapper>
        </form>
      </Card>
    </Container>
  );
};
