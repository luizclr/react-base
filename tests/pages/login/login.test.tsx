import { faker } from "@faker-js/faker";
import { fireEvent, screen, waitFor } from "@testing-library/react";
import { StyleGuide } from "react-styled-guide";

import { Login } from "~/pages/login/login.container";
import initialStyleGuide from "~/style-guide";

import { render } from "#/test-utils/render";

const makeSut = (): {
  emailInput: HTMLInputElement;
  emailError: HTMLElement;
  passInput: HTMLInputElement;
  passError: HTMLElement;
  submitButton: HTMLElement;
  styleGuide: StyleGuide;
} => {
  const { getByRole, getByTestId } = render(<Login />);

  const submitButton = getByRole("button");
  const emailInput = getByTestId("email-input") as HTMLInputElement;
  const emailError = getByTestId("email-error");
  const passInput = getByTestId("password-input") as HTMLInputElement;
  const passError = getByTestId("password-error");

  const styleGuide = initialStyleGuide.styleGuide;

  return { emailInput, emailError, passInput, passError, submitButton, styleGuide };
};

describe("<Login />", () => {
  it("should not show errors on component's first render", async () => {
    // given
    const { emailError, passError } = makeSut();

    // then
    expect(emailError.innerHTML).toEqual("");
    expect(passError.innerHTML).toEqual("");
  });

  it("should show errors when values are invalid", async () => {
    // given
    const { emailInput, emailError, passInput, passError, submitButton } = makeSut();
    const emailErrorMessage = "Invalid email";
    const emailPassMessage = "Password must contain at least 8 characters";

    // when
    fireEvent.change(emailInput, { target: { value: faker.word.sample() } });
    fireEvent.change(passInput, { target: { value: faker.string.alphanumeric(7) } });

    await waitFor(() => fireEvent.submit(submitButton));
    // then

    expect(await screen.findAllByRole("form-field-error")).toHaveLength(2);
    expect(emailError.innerHTML).toEqual(emailErrorMessage);
    expect(passError.innerHTML).toEqual(emailPassMessage);
  });

  it("should show errors when values are valid", async () => {
    // given
    const { emailInput, emailError, passInput, passError, submitButton } = makeSut();

    // when
    fireEvent.change(emailInput, { target: { value: faker.internet.email() } });
    fireEvent.change(passInput, { target: { value: faker.string.alphanumeric(8) } });

    await waitFor(() => fireEvent.submit(submitButton));
    // then

    expect(await screen.findAllByRole("form-field-error")).toHaveLength(2);
    expect(emailError.innerHTML).toEqual("");
    expect(passError.innerHTML).toEqual("");
  });

  it("should be able to mount <Login /> component", async () => {
    // given
    const { asFragment } = render(<Login />);

    // when
    const sut = asFragment();

    // then
    expect(sut).toMatchSnapshot();
  });
});
