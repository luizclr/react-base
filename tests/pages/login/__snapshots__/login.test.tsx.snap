// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should be able to mount <Login /> component 1`] = `
<DocumentFragment>
  .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: inherit;
}

.c5 {
  padding: 8px;
  border-radius: 4px;
  border: 1px solid #bababa;
  outline: none;
  margin-top: 4px;
}

.c5:focus {
  border-color: #545454;
}

.c4 {
  font-size: 12pt;
  color: #707070;
}

.c6 {
  font-size: 10pt;
  color: #ec6666;
  margin-top: 4px;
  min-height: 16px;
}

.c0 {
  background-color: #77dd77;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  padding: 16px;
  color: #545454;
}

.c1 {
  background-color: #fff;
  border-radius: 8px;
  padding: 24px;
  max-width: 100%;
  width: 400px;
}

.c2 {
  font-size: 18pt;
  margin-bottom: 24px;
  color: #545454;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: inherit;
}

.c8 {
  background-color: #f26419;
  border-radius: 4px;
  padding: 8px;
  border: none;
  color: #fff;
  font-weight: 600;
  margin: 0 auto;
  margin-top: 8px;
  width: 50%;
}

.c8:hover {
  background-color: #c25014;
  cursor: pointer;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <form>
        <h2
          class="c2"
        >
          Login
        </h2>
        <div
          class="c3"
        >
          <label
            class="c4"
            data-testid="email-label"
            for="email"
          >
            E-mail
          </label>
          <input
            class="c5"
            data-testid="email-input"
            name="email"
            placeholder="example@example.com"
            type="email"
          />
          <p
            class="c6"
            data-testid="email-error"
            role="form-field-error"
          />
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            data-testid="password-label"
            for="password"
          >
            Password
          </label>
          <input
            class="c5"
            data-testid="password-input"
            name="password"
            placeholder="**********"
            type="password"
          />
          <p
            class="c6"
            data-testid="password-error"
            role="form-field-error"
          />
        </div>
        <div
          class="c7"
        >
          <button
            class="c8"
            data-testid="submit"
            type="submit"
          >
            SIGN IN
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
